File: func6.js
// func6

// Semigroup: An associative type with a concat method. 

const Sum = x => ({
    x,
    concat: ({x: y}) => Sum(x + y),
    inspect: () => `Sum(${x})`
})

// const result = Sum(1).concat(Sum(2)) // want Sum(3)

const All = x => ({
    x,
    concat: ({x: y}) => All(x && y),
    inspect: () => `All(${x})`
})


// const result = All(true).concat(All(true)) // All(false)


const First = x => ({
    x,
    concat: () => First(x),
    inspect: () => `First(${x})`
})


const result = First("blah").concat(First("ice cream")) // First(false)

console.log(result)

File: func6.py
# func6



class Sum:
    def __init__(self, x):
        self.x = x
        self.concat = lambda o: Sum(x + o.x)
    
    def __repr__(self):
        return 'Sum({})'.format(self.x)


# result = Sum(1).concat(Sum(2))


class All:
    def __init__(self, x):
        self.x = x
        self.concat = lambda o: All(x & o.x)
    
    def __repr__(self):
        return 'All({})'.format(self.x)


# result = All(True).concat(All(True))

class First:
    def __init__(self, x):
        self.x = x
        self.concat = lambda o: First(x)
    
    def __repr__(self):
        return 'First({})'.format(self.x)


result = First('blah').concat(First('ice cream'))

print(result)